### PR TITLE
Qualcomm AI Engine Direct - Support simple_eval in calibration, perpl…

### PR DESCRIPTION
**Pull request by winskuo-quic**
### Summary

- Enable Perplexity Evaluation on device with `llama.py`
- Evaluate perplexity after qdq cpu
- Enable quantization to use simple_eval as calibration dataset.
- Enable UT to check perplexity for QWEN, which should be more reliable than checking the string output.

Will have a follow up PR to address:
- External CI enablement for qwen on x86 (If it does not take too long).
- Hide Logits scale/offset to metadata in model

#### Script
`python examples/qualcomm/oss_scripts/llama/llama.py -b build-android -s $DEVICE -m SM8750 --prompt "What is 1+1?" --temperature 0 --model_mode kv --max_seq_len 1024 --ptq 16a8w --decoder_model qwen2_5 --eval_perplexity --tasks wikitext`

### Test plan
`python backends/qualcomm/tests/test_qnn_delegate.py -k TestExampleLLMScript.test_static_qwen2_5 --model SM8650 --build_folder build-android/ --executorch_root . -s $DEVICE`

Author: @shewu-quic, @winskuo-quic

### CODE REVIEW

**Comment by winskuo-quic** [comment]
Hi @cccclai,
We are interested in using a more robust way to evaluate the accuracy of QNN model.
Before, we only ensure the first few characters matches the golden, is rather unsafe and unstable.
This PR will evaluate e2e flow for the 3 phases: prepare_pt2e (cpu fp32), convert_pt2e(cpu qdq), QNN on device.
At the end, we use QNN ppl score to check if there's regression.
We would like to add this test to ExecuTorch's CI, using x86 emulator. However, this could probably take a couple hours (estimate 3-6 hours) to finish with x86 emulator. Do you know if there are any concerns with the CI time limit?
Thanks

**Comment by cccclai** [comment]
> Hi @cccclai, We are interested in using a more robust way to evaluate the accuracy of QNN model. Before, we only ensure the first few characters matches the golden, is rather unsafe and unstable. This PR will evaluate e2e flow for the 3 phases: prepare_pt2e (cpu fp32), convert_pt2e(cpu qdq), QNN on device. At the end, we use QNN ppl score to check if there's regression. We would like to add this test to ExecuTorch's CI, using x86 emulator. However, this could probably take a couple hours (estimate 3-6 hours) to finish with x86 emulator. Do you know if there are any concerns with the CI time limit? Thanks

There are some jobs running less frequently like https://github.com/pytorch/executorch/blob/b6b7a16df5e7852d976d8c34c8a7e9a1b6f7d005/.github/workflows/periodic.yml and https://github.com/pytorch/executorch/blob/b6b7a16df5e7852d976d8c34c8a7e9a1b6f7d005/.github/workflows/nightly.yml they won't block CI. Maybe we can use these jobs

**Comment by winskuo-quic** [comment]
> > Hi @cccclai, We are interested in using a more robust way to evaluate the accuracy of QNN model. Before, we only ensure the first few characters matches the golden, is rather unsafe and unstable. This PR will evaluate e2e flow for the 3 phases: prepare_pt2e (cpu fp32), convert_pt2e(cpu qdq), QNN on device. At the end, we use QNN ppl score to check if there's regression. We would like to add this test to ExecuTorch's CI, using x86 emulator. However, this could probably take a couple hours (estimate 3-6 hours) to finish with x86 emulator. Do you know if there are any concerns with the CI time limit? Thanks
> 
> There are some jobs running less frequently like https://github.com/pytorch/executorch/blob/b6b7a16df5e7852d976d8c34c8a7e9a1b6f7d005/.github/workflows/periodic.yml and https://github.com/pytorch/executorch/blob/b6b7a16df5e7852d976d8c34c8a7e9a1b6f7d005/.github/workflows/nightly.yml they won't block CI. Maybe we can use these jobs

Sure! We will support the CI in the future PR under these yml files. Thanks

**Comment by haowhsu-quic** [comment]
Hi @cccclai,

There are some incoming PRs would like to let you know:
- 16bit kv cache IO in runtime
- documentation of llam3_1B/3B, qwen2_0.5B/1.5B e2e with instructions & metrics (we've verified it and get pretty good result)
- Model enablement for qwen3_0.6B/1.7B, smollm2_135m_instruct, phi4_mini_instruct, gemma3. (Accuracy will be addressed in other PRs)

Thank you.

**Comment by winskuo-quic** [comment]
Hi @cccclai,
This PR is now ready to review. Thanks!
For this PR, with the following script, we are expecting to get ppl score of 19 for QNN on device, which is slightly higher than prepare_pt2e and convert_pt2e (both around 12). With this PR, you should be able to see some reasonable outputs but still not the best. The token rate should be somewhere around 125-130 tok/sec on SM8750.
`python examples/qualcomm/oss_scripts/llama/llama.py -b build-android -s $DEVICE -m SM8750 --prompt "What is 1+1?" --temperature 0 --model_mode kv --max_seq_len 1024 --ptq 16a8w --decoder_model qwen2_5 --eval_perplexity --tasks wikitext --limit 1`

After this PR is merged, we will push another PR that applies some optimization. With those optimizations, we should be able to get ppl score of 12 for QNN on device, which aligns with prepare_pt2e and convert_pt2e.
Thanks.

**Review by cccclai — APPROVED** [review]
Looks great, thank you!
